### DISCUSSION TITLE
How to change the default C compiler with my custom compiler?

### DISCUSSION THREAD

**Discussion by Ahmed-Bilal-01** [eclipse-cdt-cloud/cdt-cloud#39] [Q&A]
Hi, actually I am working on a embedded problem for which I have developed my own custom C compiler. Now I want to integrate it in CDT IDE instead of default GCC compiler for C files.
Please give me full step by step guidance how to do it? like in which files (with path flow) I need to make changes. How will I be able to add my custom compiler's path etc.  ( I have tested my compiler, it works fine in compilation of c files)?

**Comment by planger** [comment]
Hi,

Thank you for your interest in CDT Cloud!

What exactly do you mean by "integrate it in CDT IDE"? If you mean that you'd like to invoke your compiler to build your project in CDT Cloud Blueprint, the easiest way would be to add a `.theia/tasks.json` in your workspace folder and add a build task that invokes your compiler.

```json
{
  // See https://go.microsoft.com/fwlink/?LinkId=733558
  // for the documentation about the tasks.json format
  "version": "2.0.0",
  "tasks": [
    {
      "type": "shell",
      "command": "<the-command-to-build-your-project>",
      "group": {
        "kind": "build",
        "isDefault": true
      }
    }
  ]
}
```

With the `tasks.json` in your workspace, the build command of CDT Cloud Blueprint will invoke `<the-command-to-build-your-project>`. For more information what you can configure in the `tasks.json`, please refer to the [VS Code documentation](https://code.visualstudio.com/docs/editor/tasks), as it is inline with what Theia provides.

I'm sorry if I misunderstood your question. In that case, please provide more context. Thank you!

**Reply by Ahmed-Bilal-01** [reply]
By integration I mean, when I execute any C file instead of GCC compile my custom compiler compile it. Like in photo below i want to replace gcc with customCompiler. For that I want to know how to do it.

![gcc](https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/137169507/f3a6cec6-5fde-424f-a7f9-af9dd8e0edd1)


Secondly, I can't find .theia/tasks.json file in CDT Cloud source code:
![noTasksJson](https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/137169507/57cdad3e-4559-4131-94b7-bc41549ff90c)

**Reply by planger** [reply]
What task or command did you execute to get this output?
The `.theia/tasks.json` or `.vscode/tasks.json` need to be in the workspace folder that is opened in CDT Cloud Blueprint. In this file you define the build task of the workspace. When the user then invokes a build, CDT Cloud Blueprint (Theia) will look into `.theia/tasks.json` or `.vscode/tasks.json` and find the default build task and run it.

**Comment by Ahmed-Bilal-01** [comment]
@planger ,I think you confused my problem let me explain again.

I have installed extension CodeRunner in CDT to run C file code. when I run code it uses gcc compiler for compilation of C files (it is because gcc is default integrated compiler in CDT for C files) what I want is that instead of that gcc compiler when I run C file in CDT it compiles with my custom compiler assume my compiler name customCompilerC.

![gcc](https://github.com/eclipse-cdt-cloud/cdt-cloud/assets/137169507/f13a35fc-01e4-4583-bb34-cc99db33f409)

as in this picture there is command that is executed with gcc compiler, I want it to be like:

[Running] cd "/home/ahmed/Desktop/" && **customCompilerC** hello22.c -o hello22 && "/home/ahmed/Desktop/"hello22

for this change what I think is to do changes in source code where C compiler is set to gcc in CDT source code. For that I need help how to change it like e.g setting my customCompilerC path etc. so that CDT can recognize my compiler.

**Reply by planger** [reply]
Thank you for giving more context. The fact that you use *CodeRunner* to run the build was so far unknown.

CDT Cloud doesn't specify what the default compiler is. The reason `gcc` is used in your case seems to be that this is defined by *CodeRunner* by default: https://github.com/formulahendry/vscode-code-runner/blob/master/package.json#L137


But it looks like you can [configure that in *CodeRunner*](https://github.com/formulahendry/vscode-code-runner/tree/master?tab=readme-ov-file#configuration).

**Comment by Ahmed-Bilal-01** [comment]
Thanks for clarity,
@planger , In that case I want to know how can I change menu Toolbar?
Like I want to do some changes in Run option menu. I want to add one more option in Run menu options, and clicking on it I want to execute a bash command

Please guide me on it, how can I do this through source code of CDT.

**Reply by planger** [reply]
You'd need to create a Theia Extension, see [the Theia docs](https://theia-ide.org/docs/authoring_extensions/) for more details.
Once you have that extension, you can follow the Theia documentation also for creating a [command and a menu](https://theia-ide.org/docs/commands_keybindings). I think the menu path of *Run* is `DebugMenus.DEBUG`.
If you want to execute a bash command, probably the most straightforward way is to run a task (see [Theia Doc on tasks](https://theia-ide.org/docs/tasks/)).

**Comment by Ahmed-Bilal-01** [comment]
Hi @planger,

I hope you are doing well.

After doing extensive research I came to know that Eclipse IDE for Embedded C/C++ Developers an be the right solution for me. But I am unable to find its repo on GitHub.

I will be very thankful to you if you can provide me link to the specific repo of Eclipse IDE for Embedded C/C++ Developers.

**Reply by planger** [reply]
Hi,

I think the source code is located in the repositories of the [eclipse-embed-cdt Github org](https://github.com/eclipse-embed-cdt/).

Best wishes,

Philip

**Comment by Ahmed-Bilal-01** [comment]
@planger ,

Thanks for sharing it. I am confused about how to build source code from there.

Like there are different multiple sub repos, which one should I clone as I need full Embedded C/C++ Developers features? and there is no proper documentation regarding how to build and run IDE app.

**Reply by planger** [reply]
It is probably best if you contact the maintainers of embedded C/C++ dev packages for those questions:
https://eclipse-embed-cdt.github.io/support/

**Comment by Ahmed-Bilal-01** [comment]
@planger ,

I want to know by any chance Theia or CDT Blueprint IDE's can support debugging for embedded family Cortex R5 by ARM family? if yes how can I do that?

**Reply by planger** [reply]
Yes, with the right VS Code extensions this should work.

There are a few options you can explore and play around with:
https://open-vsx.org/extension/marus25/cortex-debug
https://open-vsx.org/extension/eclipse-cdt/cdt-gdb-vscode
https://open-vsx.org/extension/webfreak/debug
